### PR TITLE
Add refresh token cookie path to client

### PR DESCRIPTION
**Pull request by tynyttie**
## Summary
Add refresh token. cookie path  to client_config.  This path will be added to the refresh_token cookie when cookies are generated. If no cookie config is present default to the sign in service refresh path.

## Related issue(s)
https://github.com/department-of-veterans-affairs/va.gov-team/issues/55778

[ZinHub: ticket 55778](https://app.zenhub.com/workspaces/identity-5f5bab705a94c9001ba33734/issues/gh/department-of-veterans-affairs/va.gov-team/55778)

[](url)
## Testing done

Testing Steps:
 Download [sign-in-service-client-sinatra](https://github.com/department-of-veterans-affairs/sign-in-service-client-sinatra/tree/main) from gitlab. 
 
1. In Sinatra root file structure, Open rails console.
2.  Run. config.update(update(refresh_token_path:'/test_url')) 
3. Start server - rails s
4. Start sign-in-service-client-sinatra - bin/server
5. login with test user
6. Checked Respone_header - "set-cookie" to see that path: /test_url was added. 

## Screenshots
Screenshot 2023-06-27 at 10 06 14 AM (2)

## What areas of the site does it impact?
Sign-In Service

### CODE REVIEW

File: app/models/sign_in/client_config.rb [hunk 1]
```diff
@@ -33,6 +33,10 @@ def cookie_auth?
       authentication == Constants::Auth::COOKIE
     end
 
+    def refresh_token_path
+      Constants::Auth::REFRESH_ROUTE_PATH
+    end
```
Comment (rileyanderson) [inline app/models/sign_in/client_config.rb:38]: This is overriding the column method. Whenever you call `config.refresh_token_path` it will always call this method which returns the default and not what's in the database. You should avoid overriding column methods as it causes unexpected behavior.

File: config/settings/development.yml [hunk 2]
```diff
@@ -9,8 +9,8 @@ betamocks:
   enabled: true
   recording: false
   # the cache dir depends on how you run the api
-  # cache_dir: ../vets-api-mockdata # via rails; e.g. bundle exec rails s or bundle exec rails c
-  cache_dir: /cache # via docker; e.g. make up or make console
+  cache_dir: ../vets-api-mockdata # via rails; e.g. bundle exec rails s or bundle exec rails c
+  # cache_dir: /cache # via docker; e.g. make up or make console
```
Comment (rileyanderson) [inline config/settings/development.yml:13]: this shouldn't be changed

File: db/schema.rb
Comment (rileyanderson) [inline db/schema.rb:1]: This needs to be in a separate PR. All database changes (schema, and migrations) need to be merged and migrated before app code changes

Also, you are missing the migration file.

File: spec/models/sign_in/client_config_spec.rb [hunk 3]
```diff
@@ -319,4 +319,14 @@
...
   end
+
+  describe 'refresh_token_path' do
+    subject { client_config.refresh_token_path }
+
+    let(:result) { '/v0/sign_in/refresh' }
+
+    it 'returns the refresh route path' do
+      expect(subject).to eq(result)
+    end
```
Comment (rileyanderson) [inline spec/models/sign_in/client_config_spec.rb:330]: this isn't testing much. You need to test all cases.

**Review by rileyanderson — CHANGES_REQUESTED** [review]
- This does not work as expected. It always returns the default refresh route path.
- Schema changes and migration files need to be in a separate PR and merged and migrated before app code changes are made.
  - You are missing your migration file.
- You need to add better tests that test all cases and scenarios

**Review by rileyanderson — CHANGES_REQUESTED** [review]

File: app/services/sign_in/token_serializer.rb [hunk 4]
```diff
@@ -119,5 +119,9 @@ def anti_csrf_token
     def client_config
       @client_config ||= session_container.client_config
     end
+
+    def refresh_token
```
Comment (rileyanderson) [inline app/services/sign_in/token_serializer.rb:123]: you should add `path` to this method name so it doesn't get confused with being the refresh token

File: spec/services/sign_in/token_serializer_spec.rb [hunk 5]
```diff
@@ -116,16 +116,19 @@
         end
       end
 
-      context 'and client is not configured to check for anti csrf' do
-        let(:anti_csrf) { false }
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:120]: why are you removing this?

File: spec/services/sign_in/token_serializer_spec.rb [hunk 6]
```diff
@@ -116,16 +116,19 @@
...
+      context 'when client_config has refresh_token_path' do
+        let(:client_config) { create(:client_config, authentication:, refresh_token_path: refresh_path) }
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:120]: you should use the `client_config` that gets created at the begging of the test. modify it to include `refresh_token_path` then in your tests you override `refresh_token_path` instead of the entire `client_config`

File: spec/services/sign_in/token_serializer_spec.rb [hunk 7]
```diff
@@ -116,16 +116,19 @@
...
 
-        it 'does not anti csrf token cookie' do
-          subject
-          expect(cookies[anti_csrf_token_cookie_name]).to eq(nil)
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:124]: why are you removing this?

File: spec/services/sign_in/token_serializer_spec.rb [hunk 8]
```diff
@@ -116,16 +116,19 @@
...
+        it 'returns the refresh_token_path from client_config' do
+          expect(client_config.refresh_token_path).to eq(refresh_path)
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:123]: you should be testing the `token_serializer` here not the `client_config`.

File: spec/services/sign_in/token_serializer_spec.rb [hunk 9]
```diff
@@ -116,16 +116,19 @@
...
         end
+      end
 
-        it 'returns an empty hash' do
-          expect(subject).to eq({})
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:128]: why are you removing this?

File: spec/services/sign_in/token_serializer_spec.rb [hunk 10]
```diff
@@ -116,16 +116,19 @@
...
+      context 'when client_config does not have refresh_token_path' do
+        let(:client_config) { create(:client_config, authentication:, refresh_token_path: '') }
+
+        it 'returns nil when refresh_token is not present' do
+          expect(client_config.refresh_token_path).to eq('')
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:131]: same comments as above. you should be testing the `token_serializer` when `client_config.refresh_path` is `nil`. also, `''` is a string not `nil`

**Review by rileyanderson — CHANGES_REQUESTED** [review]

File: spec/services/sign_in/token_serializer_spec.rb (hunk 9 above)
Comment (tynyttie) [reply spec/services/sign_in/token_serializer_spec.rb:128]: This was removed by mistake.

File: spec/services/sign_in/token_serializer_spec.rb (hunk 5 above)
Comment (tynyttie) [reply spec/services/sign_in/token_serializer_spec.rb:120]: This was removed by mistake.

File: app/services/sign_in/token_serializer.rb [hunk 11]
```diff
@@ -119,5 +119,13 @@ def anti_csrf_token
     def client_config
       @client_config ||= session_container.client_config
     end
+
+    def refresh_token_path
+      @refresh_token_path = if client_config.refresh_token_path.nil?
+                              Constants::Auth::REFRESH_ROUTE_PATH
+                            else
+                              client_config.refresh_token_path
+                            end
```
Comment (rileyanderson) [inline app/services/sign_in/token_serializer.rb:128]: since this is only used once you don't need to set an instance variable. You can keep it how you had before but with the addition of `path` in the method name. 
```ruby
def refresh_token_path
  client_config.refresh_token_path || Constants::Auth::REFRESH_ROUTE_PATH
end
```

File: spec/services/sign_in/token_serializer_spec.rb [hunk 12]
```diff
@@ -128,6 +128,21 @@
           expect(subject).to eq({})
         end
       end
+
+      context 'when client_config has refresh_token_path' do
+        it 'returns the refresh_token_path from client_config' do
+          subject
+          expect(cookies[refresh_token_cookie_name][:path]).to eq(refresh_path)
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:135]: you need to set `refresh_path` to something besides the default here. This needs to test that if the `client_config.refresh_token_path` is set to something besides the default it will return that path in the cookie.

File: spec/services/sign_in/token_serializer_spec.rb [hunk 13]
```diff
@@ -128,6 +128,21 @@
...
+        end
+      end
+
+      context 'when client_config does not have refresh_token_path' do
+        it 'returns nil when refresh_token is not present' do
+          subject
+          cookies[refresh_token_cookie_name] = { path: nil }
+          expect(cookies[refresh_token_cookie_name][:path]).to eq(nil)
+        end
+      end
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:145]: This is not what is expected. When the `refresh_token_path` is `nil` it should return the default.

**Review by rileyanderson — CHANGES_REQUESTED** [review]

File: spec/services/sign_in/token_serializer_spec.rb [hunk 14]
```diff
@@ -128,6 +128,30 @@
...
+      context 'when client_config does not have refresh_token_path' do
+        before do
+          client_config.refresh_token_path = nil
+        end
+
+        it 'returns default refresh_token when path is nil' do
+          subject
+          expect(cookies[refresh_token_cookie_name][:path]).to eq(refresh_path)
+        end
+      end
```
Comment (rileyanderson) [inline spec/services/sign_in/token_serializer_spec.rb:154]: Up near the top of this file there is a `client_config` that is defined:
```ruby
let(:client_config) { create(:client_config, authentication:, anti_csrf:) }
```

You should add `refresh_token_path` to this and define a `refresh_token_path`. 
Then in your tests you modify `refresh_token_path` to what you want. You shouldn't be modifying test objects like this: `client_config.refresh_token_path = nil`

You should also modify the `expected_refresh_token_cookie` to work with your newly defined `refresh_token_path`. Then, in your assertions, you should check that the entire cookie equals the expected cookie, not just the path.

**Review by rileyanderson — CHANGES_REQUESTED** [review]

File: spec/services/sign_in/token_serializer_spec.rb (hunk 14 above)
Comment (tynyttie) [reply spec/services/sign_in/token_serializer_spec.rb:154]: @rileyanderson  line 97 to 100 is testing the testing the entire cookie. Is there a need to repeat this? I do see a need to replace refresh_path in "expected_refresh_token_cookie"  with refresh_token_path that's added to 
let(:client_config) { create(:client_config, authentication:, anti_csrf:, refresh_token_path:) }

97  it 'sets refresh token cookie' do
        subject
        expect(cookies[refresh_token_cookie_name]).to eq(expected_refresh_token_cookie)
100     end
